### PR TITLE
Add query support for delegated identity entity types

### PR DESCRIPTION
**Pull request by AndriySvyryd**
* Add QueryCompilationContext.QuerySourceEntityTypeMapping to keep track of the mapping from IQuerySource to IEntityType
* Change EntityQueryModelVisitor.IterateCompositePropertyExpression to propagate the IEntityType through navigations and rename it to MemberAccessBindingExpressionVisitor.GetPropertyPath

### CODE REVIEW

File: src/EFCore.Relational/Query/RelationalQueryModelVisitor.cs [hunk 1]
```diff
@@ -1597,16 +1597,15 @@ var referencedQuerySource
...
+                QueryCompilationContext.AddOrUpdateMapping(mapping.Key, mapping.Value);
             }
 
             var innerItemParameter
                 = Expression.Parameter(
                     innerShaper.Type,
                     joinClause.ItemName);
 
-            AddOrUpdateMapping(joinClause, innerItemParameter);
+            QueryCompilationContext.AddOrUpdateMapping(joinClause, innerItemParameter);
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/RelationalQueryModelVisitor.cs]: Why this was changed? `AddOrUpdatingMapping` does the same thing. Either we keep the function and use it in all cases (including 1 above) or we remove the function.

Comment (AndriySvyryd) [inline src/EFCore.Relational/Query/RelationalQueryModelVisitor.cs]: I removed the function as it seemed redundant

File: src/EFCore/Query/EntityQueryModelVisitor.cs [hunk 2]
```diff
@@ -752,7 +756,8 @@ public override void VisitQueryModel([NotNull] QueryModel queryModel)
                     _expression.Type.GetSequenceType(),
                     fromClause.ItemName);
 
-            AddOrUpdateMapping(fromClause, CurrentParameter);
+            Expression expression = CurrentParameter;
```
Comment (smitpatel) [inline src/EFCore/Query/EntityQueryModelVisitor.cs]: pass directly to function?

Comment (AndriySvyryd) [inline src/EFCore/Query/EntityQueryModelVisitor.cs]: Refactoring gone wrong

File: src/EFCore/Query/ExpressionVisitors/Internal/QuerySourceTypeExpressionVisitor.cs [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+            }
+        }
+
+        /// <summary>
+        ///     This API supports the Entity Framework Core infrastructure and is not intended to be used
+        ///     directly from your code. This API may change or be removed in future releases.
+        /// </summary>
+        protected override Expression VisitConstant(ConstantExpression constantExpression)
+        {
+            if (constantExpression.Type.GetTypeInfo().IsGenericType
```
Comment (smitpatel) [inline src/EFCore/Query/ExpressionVisitors/Internal/QuerySourceTypeExpressionVisitor.cs]: you can use `constantExpression.IsEntityQueryable` extension method.

**Comment by anpete** [comment]
@AndriySvyryd Can we go over the changes in person tomorrow with the query team? :trollface:

**Comment by anpete** [comment]
@AndriySvyryd @smitpatel and I think we have fixed the issue (#8105) where non-qsres were showing up in binding, hopefully you can simplify things a lot after this fix goes in.

**Comment by anpete** [comment]
@AndriySvyryd If that works, we speculate that the simplest way to bind entity types will be to maintain a QuerySource -> IEntityType map.

**Comment by AndriySvyryd** [comment]
@anpete The fix worked. I am making the changes and adding tests as discussed.

**Comment by AndriySvyryd** [comment]
Updated. I'll send out another PR with functional tests to keep this PR size down.

**Comment by anpete** [comment]
@AndriySvyryd It seems like we should be able to avoid the change to EntityQueryable because we can just build the QuerySource -> EntityType map as we bind through navs during nav rewrite. Thoughts?

**Comment by AndriySvyryd** [comment]
@anpete In my next PR I found that `EntityEntry.GetDatabaseValues()` for an owned entity requires the query root to be the owned entity type. So we need to put the EntityType in the query expression to be able to generate a different cache key.

**Comment by anpete** [comment]
@AndriySvyryd Can we tweak the query for GetDatabaseValues to use the nav.? E.g. customers.Where(c => c.CustomerID == "ALFKI").Select(c => c.Address)

**Comment by AndriySvyryd** [comment]
@anpete Updated

File: src/EFCore.InMemory/Query/ExpressionVisitors/Internal/InMemoryEntityQueryableExpressionVisitor.cs [hunk 4]
```diff
@@ -51,7 +51,10 @@ protected override Expression VisitEntityQueryable(Type elementType)
         {
             Check.NotNull(elementType, nameof(elementType));
 
-            var entityType = _model.FindEntityType(elementType);
+            if (!QueryModelVisitor.QueryCompilationContext.QuerySourceEntityTypeMapping.TryGetValue(_querySource, out var entityType))
```
Comment (anpete) [inline src/EFCore.InMemory/Query/ExpressionVisitors/Internal/InMemoryEntityQueryableExpressionVisitor.cs]: DRY to method on QueryCompilationContext

File: src/EFCore/Query/QueryCompilationContext.cs [hunk 5]
```diff
@@ -107,6 +107,14 @@ public class QueryCompilationContext
         public virtual QuerySourceMapping QuerySourceMapping { get; } = new QuerySourceMapping();
 
         /// <summary>
+        ///     Gets the mapping to query source
+        /// </summary>
+        /// <value>
+        ///     The mapping to query source
+        /// </value>
+        public virtual Dictionary<IQuerySource, IEntityType> QuerySourceEntityTypeMapping { get; } = new Dictionary<IQuerySource, IEntityType>();
```
Comment (anpete) [inline src/EFCore/Query/QueryCompilationContext.cs]: Can we keep the dictionary private and add accessor methods.

File: src/EFCore/Query/ResultOperators/Internal/IncludeResultOperator.cs [hunk 6]
```diff
@@ -94,7 +103,51 @@ public virtual void AppendToNavigationPath([NotNull] IReadOnlyList<PropertyInfo>
         ///     This API supports the Entity Framework Core infrastructure and is not intended to be used
         ///     directly from your code. This API may change or be removed in future releases.
         /// </summary>
-        public override string ToString() 
+        public virtual INavigation[] GetNavigationPath([NotNull] QueryCompilationContext queryCompilationContext)
```
Comment (anpete) [inline src/EFCore/Query/ResultOperators/Internal/IncludeResultOperator.cs]: Should have told you to ignore the old Include code paths. It is all going away :trollface:

**Comment by anpete** [comment]
Looks good, minor comments.

**Comment by anpete** [comment]
cc @maumar for navs review.

**Pull request merged**